x/staking/src/keeper/historical_info.rs

```diff
@@ -0,0 +1,96 @@
+use gears::store::database::ext::UnwrapCorrupt;
+
+use super::*;
+use crate::{historical_info_key, HistoricalInfo};
+
+impl<
+        SK: StoreKey,
+        PSK: ParamsSubspaceKey,
+        AK: AccountKeeper<SK>,
+        BK: BankKeeper<SK>,
+        KH: KeeperHooks<SK>,
+    > Keeper<SK, PSK, AK, BK, KH>
+{
+    pub fn track_historical_info<DB: Database>(&self, ctx: &mut BlockContext<'_, DB, SK>) {
+        let params = self.staking_params_keeper.get(ctx);
+        let entry_num = params.historical_entries;
+
+        // Prune store to ensure we only have parameter-defined historical entries.
+        // In most cases, this will involve removing a single historical entry.
+        // In the rare scenario when the historical entries gets reduced to a lower value k'
+        // from the original value k. k - k' entries must be deleted from the store.
+        // Since the entries to be deleted are always in a continuous range, we can iterate
+        // over the historical entries starting from the most recent version to be pruned
+        // and then return at the first empty entry.
+
+        if (ctx.height() as i64 - entry_num as i64) >= 0 {
+            for i in (ctx.height() - entry_num as u64)..=0 {
+                // SAFETY: the BlockContext is infallible context in context of StoreErrors
+                if let Some(_info) = self.historical_info(ctx, i).unwrap() {
```

Comment:
@NYBACHOK I think we should have a `GasError` on the store method and have add an `unwrap_gas` method to `Result`

Reply:
What it gives for us? Unified unwrap in case of infallible ctx?